### ISSUE TITLE
With a custom_forward function, n_steps != 1 alters input tensor shapes, leading to shape mismatches in the model pass. Am I missing something?

### ISSUE THREAD

**Issue by nealgravindra** [pytorch/captum#625]
**Overview**

I am using a Pytorch-Geometric Data Object to handle my input, `pg_object=Pytorch-Geometric Data object)`. Similar to [this collab notebook](https://colab.research.google.com/drive/1fLJbFPz0yMCQg81DdCP5I8jXw9LoggKO?usp=sharing), which is featured on Pytorch-Geometric's tutorial page, I am using a custom_forward wrapper function for `Captum` compatibility.

**Detailed issue**

This is my adaptation:
```
def custom_forward(x, pg_object):
    edge_index = pg_object.edge_index
    edge_attr =  utils.edge_set_reshape(pg_object).float() # have to reshape the edge attributes
    print(edge_index.shape) # print to show issue
    return model(x, edge_index, edge_attr)

ig = IntegratedGradients(custom_forward)

print(batch)
attr = ig.attribute(batch.x, target=2,
                            additional_forward_args=(batch,))
```

The output of this is a shape mismatch error, after printing these "diagnostics" :
```
Data(edge_attr=[3133, 18], edge_index=[2, 3133], x=[689, 25626], y=[689]) # this is the batch, a Pytorch-Geometric obj
torch.Size([100, 3133]) # edge_index size after passing through IntegratedGradients

```

Which shows that `edge_index` gets modified by the number of steps after passing through `IntegratedGradients`. Expected input for the edge index has shape [2, 3133] but typically, it gets multiplied to [2 * `n_steps`, 3133], which causes problems in a pass through the model. 

**Question**

Am I doing something wrong here in how I implement IntegratedGradients with a model using Pytorch-Geometric layers? I can get `attr` if I set `n_steps=1` but to me, that means the estimate of the integral will be meaningless?

**Comment by vivekmig** [comment]
Hi @nealgravindra , this behavior is surprising, only tensors in additional_forward_args should be expanded, since batch is PyTorch geometric object, it should not be modified or expanded in Integrated Gradients. I tried to reproduce the issue with code similar to your example, but it [seems to work](https://colab.research.google.com/drive/1iRuKRXc3FQzSknS63Nx5UjGSWJRR6Mjb?usp=sharing) as expected. Can you provide a Colab notebook to reproduce with the specific package versions you are using?

You can also try setting internal_batch_size to equal to the batch size (first dimension of batch.x), which would essentially process each step in the integral separately.

**Comment by nealgravindra** [comment]
Hi @vivekmig,

The notebook I am working with is [here](https://github.com/nealgravindra/sc2drug), as `get_Yprime_dev.ipynb`. This is based on follow-up work so it's a huge pain to get everything, don't feel pressure to look at this horrendous mess.

Nonetheless, I have figured out that calling a `utils.py` function, which breaks the computation graph, on `edge_attr` causes at least some problems. Having a simpler model (without a transformer) gets the boilerplate code to work.

Not sure what to think about that... but I'm at least getting by with the following code: 

```
class GAT(torch.nn.Module):
    def __init__(self):
        super(GAT, self).__init__()
        self.gat1 = data.GATConv(d.num_node_features, out_channels=8,
                                 heads=8, concat=True, negative_slope=0.2,
                                 dropout=0.4, bias=True)
        self.gat2 = data.GATConv(8*8, out_channels=d.y.unique().size()[0],
                                 heads=8, concat=False, negative_slope=0.2,
                                 dropout=0.4, bias=True)

    def forward(self, x, edge_index):
#         x, edge_index = data.x, data.edge_index
        x, attn_gat1 = self.gat1(x, edge_index)
        hprime = x # store embeddings 
        x = F.elu(x)
        x, attn_gat2 = self.gat2(x, edge_index)
        return F.log_softmax(x, dim=1)

# model = GAT_transformer_mlp_batch()
model = GAT()
model.eval()
model.load_state_dict(torch.load(kwargs['modeleval_pkl'], map_location=torch.device('cpu')))

# recraft dataset
dprime = Data(x=d.x, edge_index=d.edge_index, y=torch.cat((torch.tensor(np.arange(md.shape[0])).view(-1, 1), d.y.view(-1, 1)), dim=1))

# custom forward (does not work with utils.reshape fx)
def custom_forward(x, pg_object):
    edge_index = pg_object.edge_index
    edge_attr =  utils.edge_set_reshape(pg_object).float()
    return model(x, edge_index, edge_attr)

ig = IntegratedGradients(custom_forward)
nt = NoiseTunnel(ig)

# try on batched mode
cd_test = data.ClusterData(dprime,5000)
cl_test = data.ClusterLoader(cd_test,256,shuffle=True)

results = pd.DataFrame()
for i, batch in enumerate(cl_test):
    print('')
    print('  ', batch)
    
    attr = nt.attribute(batch.x, 
                        target=0,
                        additional_forward_args=(batch,),
                        nt_type='smoothgrad', nt_samples=6)
```

As such, you are welcome to close the comment. Thanks!

**Comment by NarineK** [comment]
@nealgravindra, @vivekmig  is this still an open issue ?

**Comment by nealgravindra** [comment]
Hey, the issue was on my end I believe, due to indexing in numpy prior to passing to the set transformer class. Thanks for all of your help! Feel free to close the issue
